x/bank/keeper/keeper.go

```diff
@@ -398,28 +373,40 @@ func (k BaseKeeper) BurnCoins(ctx sdk.Context, moduleName string, amt sdk.Coins)
 		panic(sdkerrors.Wrapf(sdkerrors.ErrUnauthorized, "module account %s does not have permissions to burn tokens", moduleName))
 	}
 
-	err := k.subUnlockedCoins(ctx, acc.GetAddress(), amt)
+	err := k.subUnlockedCoins(ctx, acc.GetAddress(), amounts)
 	if err != nil {
 		return err
 	}
 
-	// update total supply
-	supply := k.GetTotalSupply(ctx)
-	supply = supply.Sub(amt)
-
-	k.setSupply(ctx, supply)
+	var newSupply sdk.Coins
+	for _, amount := range amounts {
+		supply := k.GetSupply(ctx, amount.GetDenom())
+		supply = supply.Sub(amount)
+		newSupply = append(newSupply, supply)
+	}
+	k.setSupply(ctx, newSupply)
 
 	logger := k.Logger(ctx)
-	logger.Info("burned tokens from module account", "amount", amt.String(), "from", moduleName)
+	logger.Info("burned tokens from module account", "amount", amounts.String(), "from", moduleName)
 
 	// emit burn event
 	ctx.EventManager().EmitEvent(
-		types.NewCoinBurnEvent(acc.GetAddress(), amt),
+		types.NewCoinBurnEvent(acc.GetAddress(), amounts),
 	)
 
 	return nil
 }
 
+func (k BaseKeeper) setSupply(ctx sdk.Context, coins sdk.Coins) {
+	store := ctx.KVStore(k.storeKey)
+	supplyStore := prefix.NewStore(store, types.SupplyKey)
+
+	for i := range coins {
+		bz := k.cdc.MustMarshalBinaryBare(&coins[i])
```

Comment:
Also noticing that this doesn't seem quite right. There's no need to serialize the whole `Coin`. We just need the amount serialized in the value. The denom is already serialized in the key.

Reply:
This was the first approach I took, setSupply to get one coin, I can revert it. I agree on the coin serialization. Thanks!